### ISSUE TITLE
paddlepaddle平台下搭建softmax分类器时对输入输出层设置的一些疑问

### ISSUE THREAD

**Issue by dockiHan** [PaddlePaddle/Paddle#3730]
各位老师好，最近自己在尝试搭建一个分类器，关于softmax分类器label的输入输出层设置有一些疑问，希望各位老师解答一下：
1.根据其他平台或者文档的说明，softmax分类器的输出层应该是一个1 * classdim的向量，向量中每一个值表示对应到该类别的可能性大小。 但是我根据paddlepaddle文档搭建的softmax分类器只输出了一个浮点数，请问这个浮点数代表的是什么？是softmax之后取的最大值吗？
2.我该如何根据输出的一个浮点数确定所属的类别？
3.paddlepaddle下搭建分类器是否支持这样的输入输出：输入的label是1 * classdim的向量以三个类别为例，如[1，0，0]； 输出层也是对应的1 * classdim的向量如[0.9，0.3， 0.1]。这样我可以根据max值所在向量中的位置确定分类器预测出的类别。

**Comment by pkuyym** [comment]
@dockiHan 我尝试解答一下你的疑惑
1. ```Softmax```只是将输入进行概率化，输出的维度是由输入制定的，你的例子里面只输出了1个值，意味着你输入到```Softmax```的数据维度为1
2. ```Softmax```前面一般是一个全连接层```fc```，```fc```的神经元个数即为类别数
3. 如何确定类别，这个是由训练数据的```label```决定的，```label```用one-hot表示，比如 [0 1 0]表示样本为第二类，[1 0 0]表示样本为第一类，至于第一类、第二类的具体含义，是有你来决定的。

**Comment by dockiHan** [comment]
@pkuyym 非常感谢你的回答。但是还是存在一些疑惑：

1.关于分类器的配置我是参照paddlepaddle官方文档中图片分类的分类器配置的，其中label Y的配置如下：
`Y = paddle.layer.data(name="Y", type = paddle.data_type.integer_value(value_range=output_dim))`

output_dim可以理解为3，也就是3个类别；softmax所作用的全连接层fc的配置如下：

···
def logistic_regression(input, hide_act, hidden_dim, out_act, output_size):
    hide_lr = 1e-3
    out_lr = 1e-3
    hide_param_attr = paddle.attr.Param(learning_rate=hide_lr, gradient_clipping_threshold=100)
    out_param_attr = paddle.attr.Param(learning_rate=out_lr, gradient_clipping_threshold=100)
    hide_0 = paddle.layer.fc(input=input, size=hidden_dim, act=hide_act, param_attr=hide_param_attr)
    out_layer = paddle.layer.fc(input=hide_0, size=output_size, act=out_act, param_attr=out_param_attr)
    return out_layer

output = logistic_regression(input=X, 
                             hide_act=paddle.activation.Relu(), 
                             hidden_dim=hidden_dim, 
                             out_act=paddle.activation.Softmax(), 
                             output_size=output_dim)
···
其中输出层fc的size跟类别数目保持一致（即output_dim）这样配置完之后模型输出的预测结果只有1个浮点数

-------------------------------------------分割线-----------------------------------------------------------------

2. 如何将输入的label Y 定义为one-hot类型？
我曾尝试这样配置我的label：
`Y = paddle.layer.data(name="Y", type = paddle.data_type.dense_vector(dim=1))`
但是运行时会报错

**Comment by pkuyym** [comment]
@dockiHan 
1.  > 模型输出结果只有一个浮点数

请问你是怎么拿到模型的输出结果的？ ```out_layer = paddle.layer.fc(input=hide_0, size=output_size, act=out_act, param_attr=out_param_attr)```请double check一下output_size确实为3
2. label的正确配置为
```Y = paddle.layer.data(name="Y", type = paddle.data_type.integer_value(value_range=output_dim))```，这样```Y = paddle.layer.data(name="Y", type = paddle.data_type.dense_vector(dim=1))```是不正确的

**Comment by dockiHan** [comment]
@pkuyym 感谢您的帮助！ 我找到问题所在了。
我犯了一个比较低级的错误，之前用paddlepaddle搭建过LSTM对时间序列的回归模型，当时预测结果只需要取输出向量的最后一个值，我直接把这个回归模型的预测部分搬了过来 还没来得及做修改，所以对每次预测的结果都只取了向量的最后一个值，导致我看到的输出只有一个浮点数。。。
再次感谢您的帮助！
